Allow export default const enum, export default enum, export default …

NaridaL opened the pull request:
…declare class

Also export default namespace.

https://github.com/Microsoft/TypeScript/issues/3792#issuecomment-303526468

> @mhegazy what about export default type foo = {} ?

msftclas commented:
@NaridaL,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

NaridaL commented:
Hmm it seems to pass here, but I'm getting...?

```
PS C:\Users\aval\tsdev\TypeScript> npx gulp runtests-parallel
C:\Users\aval\tsdev\TypeScript\node_modules\gulp\bin\gulp.js
[03:01:04] Requiring external module ts-node/register
[03:01:04] Using gulpfile ~\tsdev\TypeScript\gulpfile.ts
[03:01:04] Starting 'build-rules'...
snip snip
[03:01:05] Finished 'built\local\run.js' after 232 ms
[03:01:05] Starting 'tests'...
[03:01:05] Finished 'tests' after 2.34 μs
[03:01:05] Starting 'runtests-parallel'...
Running tests with config: {"light":false,"workerCount":4,"taskConfigsFolder":"C:\\Users\\aval\\AppData\\Local\\Temp/ts-tests34","noColor":false}
node built\local\run.js
Discovering tests...
Discovered 11206 test files in 665ms.
Starting to run tests using 4 threads...
Batching initial test lists...
Suboptimal packing detected: no tests remain to be stolen. Reduce packing fraction from 0.9 to fix.
Batched into 3 groups. Unprofiled tests including tsrunner-compiler://tests/cases/compiler/exportDefaultNamespace.ts will be run first.

  [........................................................................................................................................................................]
Test worker encounted unexpected error and was forced to close:
        Message: Cannot read property 'runner' of undefined
        Stack: TypeError: Cannot read property 'runner' of undefined
    at process.<anonymous> (C:\Users\aval\tsdev\TypeScript\src\harness\parallel\worker.ts:143:29)
    at emitTwo (events.js:125:13)
    at process.emit (events.js:213:7)
    at process.emit (C:\Users\aval\tsdev\TypeScript\node_modules\source-map-support\source-map-support.js:431:21)
    at emit (internal/child_process.js:774:12)
    at _combinedTickCallback (internal/process/next_tick.js:141:11)
    at process._tickCallback (internal/process/next_tick.js:180:9)
[03:01:09] 'runtests-parallel' errored after 4.47 s
[03:01:09] Error: 2
    at formatError (C:\Users\aval\tsdev\TypeScript\node_modules\gulp\bin\gulp.js:169:10)
    at Gulp.<anonymous> (C:\Users\aval\tsdev\TypeScript\node_modules\gulp\bin\gulp.js:195:15)
    at emitOne (events.js:115:13)
    at Gulp.emit (events.js:210:7)
    at Gulp.Orchestrator._emitTaskDone (C:\Users\aval\tsdev\TypeScript\node_modules\orchestrator\index.js:264:8)
    at C:\Users\aval\tsdev\TypeScript\node_modules\orchestrator\index.js:275:23
    at finish (C:\Users\aval\tsdev\TypeScript\node_modules\orchestrator\lib\runTask.js:21:8)
    at cb (C:\Users\aval\tsdev\TypeScript\node_modules\orchestrator\lib\runTask.js:29:3)
    at failWithStatus (C:\Users\aval\tsdev\TypeScript\gulpfile.ts:680:9)
    at C:\Users\aval\tsdev\TypeScript\gulpfile.ts:697:17
    at <anonymous>
```

NaridaL commented on src/compiler/parser.ts:
```diff
@@ -1310,7 +1310,11 @@ namespace ts {
             nextToken();
```
nextTokenCanFollowDefaultKeyword doesn't really seem like an accurate name considering it's looking at multiple? Same with nextTokenCanFollowModifier. Would something like `declarationCanFollowExportDefault` be more appropriate?

andy-ms replied:
I think the name is fine since it doesn't take a `declaration` as input, it just goes to the next token and looks at its kind and *maybe* does some minimal lookahead.

k8w commented:
```typescript
export default const a = 'a';
```

Still get error in tsc 2.6.0-dev.20170921

mhegazy commented:
@andy-ms can you please review this change.. we need to make sure all LS features work as expected with this change..

andy-ms commented on src/compiler/parser.ts:
```diff
@@ -1310,7 +1310,11 @@ namespace ts {
+                token() === SyntaxKind.NamespaceKeyword ||
+                (token() === SyntaxKind.ConstKeyword && lookAhead(() => nextToken() === SyntaxKind.EnumKeyword)) ||
                 (token() === SyntaxKind.AbstractKeyword && lookAhead(nextTokenIsClassKeywordOnSameLine)) ||
+                (token() === SyntaxKind.DeclareKeyword && lookAhead(nextTokenIsClassKeywordOnSameLine)) ||
```
Also we should probably allow the old `module` keyword here as equivalent with `namespace` for consistency.

NaridaL replied:
What about ``export default declare abstract class` ?

andy-ms replied:
I think that should be supported too.

NaridaL commented:
@k8w See https://esdiscuss.org/topic/why-is-export-default-var-a-1-invalid-syntax for why that doesn't make sense.

NaridaL commented:
My turn to apologize for the delay; sorry.

I rewrote `nextTokenCanFollowDefaultKeyword`  to use a switch statement.
Fixing the transformer is unfortunately not quite so simple.

Currently the enum transformer generates something like `N = N || (N = {})`, with the latter two N's then getting transformed into `exports.N` by the module transformer. Something like `N = default || (default = {})`  doesn't make much sense and doesn't get transformed correctly anyways. In any case this pattern isn't transferable for ES6 modules output.

In general, the behavior of ts.ts:visitEnumDeclaration (and visitModuleDeclaration) seems weird. Doing the `|| {}` in the IIFE argument instead of in the initializer requires behavior conditional on the targeted module. (https://github.com/Microsoft/TypeScript/blob/master/src/compiler/transformers/ts.ts#L2613-L2619) 

The latest commit transforms as follows for ES6, with the further changes for commonjs etc being handled by the respective transformers.
```ts
export default enum A { FOO = 2 };
// should be transformed into
var A = A || {};
export default A;
(function (A) {
    A[A["FOO"] = 2] = "FOO";
})(A);


export enum A { FOO = 2 };
// should be transformed into
export var A = A || {};
(function (A) {
    A[A["FOO"] = 2] = "FOO";
})(A);



namespace ts {
	export enum A { FOO = 2 };
}
// should be transformed into
var ts = ts || {};
(function (ts) {
	var A = ts.A || {};
	ts.A = A;
	(function (A) {
    	A[A["FOO"] = 2] = "FOO";
	})(A);
})(ts);



```
It's very much possible I've missed some slight semantic difference...
This changes touches > 800 tests so I've only committed tests=/exportDefault.*/

NaridaL commented on tests/baselines/reference/exportDefaultNamespace.js:
```diff
@@ -0,0 +1,32 @@
+"use strict";
+exports.__esModule = true;
+// https://github.com/Microsoft/TypeScript/issues/3792
+var N = exports.N || {};
```
exports.N is a bit awkward here, but AFAIK should always be undefined, as things like the following aren't allowed?
```
export default class N {}
export class N {}
```

NaridaL replied:
Thanks for the feedback, I'll get started on it. Would a nameless namespace be equivalent to a "namespace expression"? In any case I see some issues when module=es6:

```ts
export default namespace {
    export const a = 2;
}
// namespace merge
export default namespace {
    export const b = 2;
}

// requires generated variable name:
var _gen0 = {}
(function (_gen0) {
    _gen0.a = 2;
})(_gen0);
(function (_gen0) {
    _gen0.b = 2;
})(_gen0);
```

As anonymous interfaces aren't allowed either currently, maybe this would be better handled as a separate issue?

rbuckton replied:
That's a fair point, and perhaps we should hold off on nameless enum/namespace for now until the team has had an opportunity to discuss. My intuition is that you would not be able to merge with a nameless default namespace.

@DanielRosenwasser, @mhegazy: any thoughts?

mhegazy commented:
I think we need to discuss the nameless declarations in the design meeting first.

NaridaL commented:
I've overwritten the previous commits with two new ones. The first one includes all new tests and changes to parser.ts. The changed files are limited to the new tests and 2 others. The second commit changes the emit in ts.ts, which results in a lot of changed test results. I've closely reviewed the changes in my new tests and looked at a couple of the others at random.

NaridaL commented on src/compiler/parser.ts:
```diff
@@ -1353,12 +1353,35 @@ namespace ts {
+                case SyntaxKind.DeclareKeyword:
+                case SyntaxKind.AsyncKeyword:
+                case SyntaxKind.TypeKeyword:
+                    return lookAhead(nextTokenIsIdentifierOrKeywordOnSameLine);
```
@rbuckton This simplified lookahead seems to be enough; see https://github.com/Microsoft/TypeScript/pull/18628/commits/5113249e4e3e9da299e9dfe804c7545dd9922b2e#diff-cc2127059a7624c4bd840e11d106cb81

NaridaL commented on src/compiler/parser.ts:
```diff
@@ -1326,12 +1326,12 @@ namespace ts {
             }
             if (token() === SyntaxKind.DefaultKeyword) {
-                return nextTokenCanFollowDefaultKeyword();
+                return nextTokenCanFollowDefaultModifier();
```
This whole parseModifiers method very confusing...

NaridaL commented on src/compiler/binder.ts:
```diff
@@ -459,8 +468,8 @@ namespace ts {
                     const local = declareSymbol(container.locals, /*parent*/ undefined, node, exportKind, symbolExcludes);
-                    local.exportSymbol = declareSymbol(container.symbol.exports, container.symbol, node, symbolFlags, symbolExcludes);
                     node.localSymbol = local;
+                    local.exportSymbol = declareSymbol(container.symbol.exports, container.symbol, node, symbolFlags, symbolExcludes);
```
I flipped these two lines so I can access node.localSymbol in `declareSymbol` so can check if two default exports do not have the same identifier. Maybe this is better handled in checker.ts?

NaridaL commented on tests/cases/project/declarationsExportDefault.json:
```diff
@@ -0,0 +1,17 @@
+        "namespace.ts",
+        "type.ts"
+    ]
+}
```
I added this project test to test the .d.ts output, is there a better way?

rbuckton replied:
We are trying to avoid adding new "projects" tests in favor test variations in compiler/conformance tests. You can set `// @declaration: true` in the compiler/conformance tests to get .d.ts output.

NaridaL commented on src/compiler/transformers/ts.ts:
```diff
@@ -2817,32 +2820,28 @@ namespace ts {
-                // declaration we do not emit a leading variable declaration. To preserve the
-                // begin/end semantics of the declararation and to properly handle exports
-                // we wrap the leading variable declaration in a `MergeDeclarationMarker`.
-                const mergeMarker = createMergeDeclarationMarker(statement);
```
MergeDeclarationMarker are unnecessary if the variable is initialized immediately. I think the logic handling those in the other modules is now dead.

alvis commented:
ping @andy-ms & @rbuckton

NaridaL commented:
If someone OKs my proposed transform above I can rebase this.

alvis commented:
Yes. There're still people, like me, waiting for it being merged.

banxi1988 commented:
If I export default a const enum, it will eventually produce undefined value.

```ts
const enum  Status{
   good = 0,
   bad = -1
}
export default Status; // this can lead undefined error. when using as `Status.good`
```

NaridaL commented:
@banxi1988 I don't follow...?

banxi1988 commented:
@NaridaL  sorry, I do actually encounter the undefined error caused by export default const enum type in my project. but Haven't reproduce it in a demo project. I'll try to reproduce it with a webpack configured project.

RyanCavanaugh commented:
@NaridaL if you can get this merged up with master we can give it a fresh review. Thanks for your patience!

NaridaL commented:
Rebased it...

In: tests/baselines/reference/transformApi/transformsCorrectly.transformAddCommentToNamespace.js

The comment is being output both on the var declaration and the module statement.

Could someone please clarify what the difference between an emit node and a synthesized node and when/why they should be used? Setting a flag or comment range is clearly the cause of the above issue, but it's not obvious where... it works fine in the tests I added.

rbuckton commented:
`EmitNode` is not an actual AST node, but rather is an extra set of properties attached to a `Node` that effects emit. These properties aren't added to a `Node` directly as it would increase the memory requirements for `Node` in the language service, which rarely needs to perform emit.

A "synthesized" node is a `Node` that was created synthetically as part of a transformation rather than created by the parser "from source". A "synthesized" node has `NodeFlags.Synthesized` in its `flags` property, and _usually_ has a `pos` and `end` of `-1` (although these can be set to actual positions if the synthesized node corresponds to an actual node.

rbuckton commented:
If you set `pos` or `end` on a synthesized Node (i.e. via the `setTextRange` function), it will use those positions for comment emit. You can disable this by setting an `EmitFlag` of `NoComments` via the `setEmitFlag` function.

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES2015ModuleES2015.js:
```diff
@@ -0,0 +1,122 @@
+ * export default enum A 1 comment
+ */
+export default const A = {};
+export default A;
```
This will be an error at runtime.

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES2015ModuleES2015.js:
```diff
@@ -0,0 +1,122 @@
+ * export enum B 1 comment
+ */
+export const B = {};
+export { B };
```
Also a possible error.

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES2015ModuleAMD.js:
```diff
@@ -0,0 +1,128 @@
+    /**
+     * export enum B 1 comment
+     */
+    const B = {};
```
`B` is never exported, despite the fact it is declared as `export enum B` in the test.

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES2015ModuleCommonJS.js:
```diff
@@ -0,0 +1,125 @@
+/**
+ * export enum B 1 comment
+ */
+const B = {};
```
`B` is never exported, despite the fact it is declared as `export enum B` in the test.

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES2015ModuleSystem.js:
```diff
@@ -0,0 +1,144 @@
+    return {
+        setters: [],
+        execute: function () {// https://github.com/Microsoft/TypeScript/issues/3792
+            exports_1("A", A = {});
```
`A` should not be exported as `"A"`

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES5ModuleAMD.js:
```diff
@@ -0,0 +1,128 @@
+    /**
+     * export enum B 1 comment
+     */
+    var B = {};
```
`B` is never exported, despite the fact it is declared as `export enum B` in the test.

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES5ModuleCommonJS.js:
```diff
@@ -0,0 +1,125 @@
+/**
+ * export enum B 1 comment
+ */
+var B = {};
```
`B` is never exported, despite the fact it is declared as `export enum B` in the test.

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES5ModuleES2015.js:
```diff
@@ -0,0 +1,122 @@
+ * export default enum A 1 comment
+ */
+export default var A = {};
+export default A;
```
This is an error at runtime.

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES5ModuleES2015.js:
```diff
@@ -0,0 +1,122 @@
+ * export enum B 1 comment
+ */
+export var B = {};
+export { B };
```
Also an error.

rbuckton commented on tests/baselines/reference/exportDefaultEnumTargetES5ModuleSystem.js:
```diff
@@ -0,0 +1,144 @@
+    return {
+        setters: [],
+        execute: function () {// https://github.com/Microsoft/TypeScript/issues/3792
+            exports_1("A", A = {});
```
`A` should not be exported as `"A"`.

rbuckton commented on tests/baselines/reference/exportDefaultNamespace.js:
```diff
@@ -0,0 +1,163 @@
+/**
+ * namespace B 1 leading comment
+ */
+var B = {};
```
`B` is never exported, despite the fact it is declared as `export enum B` in the test.

rbuckton commented on tests/cases/compiler/exportDefaultEnumTargetES2015ModuleAMD.ts:
```diff
@@ -0,0 +1,62 @@
+// https://github.com/Microsoft/TypeScript/issues/3792
+// @target: es2015
```
Our compiler runner supports test variations for the same source. Instead of multiple `nnnTargetXModuleY.ts` files you can do this instead:
```suggestion
// @target: es2015, es5
```

NaridaL replied:
That would be nice, but it doesn't look like it's in yet?

` Error: Unknown value 'es5, es2015' for compiler option 'target'.`

rbuckton commented on tests/cases/compiler/exportDefaultEnumTargetES2015ModuleAMD.ts:
```diff
@@ -0,0 +1,62 @@
+// https://github.com/Microsoft/TypeScript/issues/3792
+// @target: es2015
+// @module: amd
```
Our compiler runner supports test variations for the same source. Instead of multiple `nnnTargetXModuleY.ts` files you can do this instead:
```suggestion
// @module: amd, commonjs, system, es2015
```

rbuckton commented on tests/cases/compiler/exportDefaultEnumTargetES2015ModuleAMD.ts:
```diff
@@ -0,0 +1,62 @@
+// https://github.com/Microsoft/TypeScript/issues/3792
```
Consider adding these tests in a relevant location under `tests/cases/conformance/**` as `tests/cases/compiler` is extremely cluttered and unorganized.

rbuckton replied:
Also, consider adding `// @declaration: true` to verify .d.ts emit and `// @sourceMaps: true` to verify source-map emit.

TomasHubelbauer commented:
Sorry for a dumb question, but does this add support for `export default type`? The diff is more than 1k files so I haven't really tried to figure that out from the changes.

RyanCavanaugh commented:
Closing due to lack of activity here. Thanks for the work so far - if anyone would like to pick up on this feature, please do!

TomasHubelbauer commented:
Is there an "authoritative" issue for people to subscribe to in order to be kept in the loop on this feature eventually dropping? This PR doesn't reference any issues as far as I can see and searching for *export default* and *export default type* returns 50 pages of results.

damianobarbati commented:
@NaridaL why was this closed? Is there any chance of having `export default type {}`?
